File: day01.hs
-- https://adventofcode.com/2023/day/1

module Main where

main = getContents >>= \s -> do
    print $ solve part1 s
    print $ solve part2 s

solve f = sum . map (line f) . lines

line f s = 10 * head ints + last ints
  where ints = mapMaybe f (tails s)

part1 (c:_) | isDigit c = Just $ digitToInt c
part1 _ = Nothing

part2 s = part1 s <|> asum (match <$> numbers) where
    match (n,w) | isPrefixOf w s = Just n
    match _ = Nothing

numbers = zip [1..] $ words "one two three four five six seven eight nine"

File: day01.py
import sys

words = "one two three four five six seven eight nine".split()

sum1 = 0
sum2 = 0

for line in sys.stdin:
    nums1 = []
    nums2 = []

    for i in range(0,len(line)-1):
        if line[i] >= '1' and line[i] <= '9':
            nums1.append(int(line[i]))
            nums2.append(int(line[i]))
        else:
            for n,w in enumerate(words,1):
                if line.startswith(w,i):
                    nums2.append(n)
                    break

    sum1 += 10 * nums1[0] + nums1[-1]
    sum2 += 10 * nums2[0] + nums2[-1]

print(sum1)
print(sum2)
